addon/util.js

```diff
@@ -24,3 +24,50 @@ export const OWNER_KEY = (function() {
     return ownerSymbol;
   }
 })();
+
+export function isObject(value) {
+  return value !== null && typeof value === 'object' && value.constructor !== Date;
+}
+
+export function merge(data, updates) {
+  let changedKeys = [];
+  if (!updates) {
+    // no changes
+    return changedKeys;
+  }
+  let updatedKeys = Object.keys(updates);
+  for (let i = 0; i < updatedKeys.length; i++) {
+    let key = updatedKeys[i];
+    let newValue = updates[key];
+    if (isEqual(data[key], newValue)) {
+      // values are equal, nothing to do
+      // note, updates to objects should always result in new object or there will be nothing to update
+      continue;
+    }
+    if (data[key] == null || newValue == null) {
+      // reseting a value to null or assigning a value for first time can be handled for all cases
+      data[key] = newValue;
+      changedKeys.push(key);
```

Comment:
I prefer to handle `null` at top to remove the need to think about it in the rest of the cases. Technically, it can be dropped the way the code is written at the moment, but future changes may require special handling of `null` for some of the cases. By handling at the very top, we can let everything else not deal with `null`.

Reply:
@dnachev any reason for the `==` instead of `===` here?

Reply:
@runspired Handling `null` and `undefined` the same way. We can potentially handle `undefined` differently by deleting the property entirely from the hash, but not sure what we gain with this optimization.

Reply:
Okay, let's prefer reducing branching over extracting `null`-checking.

Also since the `null` case is the same as the default case if we do this there's no need for an unexpected `==`.